### ISSUE TITLE
Tempo throws 500s when rolling queriers

### ISSUE THREAD

**Issue by annanay25** [grafana/tempo#761]
**Describe the bug**

Tempo Query Frontend throws 500s when the querier deployment is rolled. 

**To Reproduce**
1. Start Tempo (SHA or version): `latest`
2. Perform Operations (Read/Write/Others): 

```
kubectl rollout restart deployment querier
```
and then watch the Query Frontend RED metrics.

**Expected behavior**

No errors on the query frontend. I would expect the querier to bail only after completing any in-flight requests.

**Environment:**
 - Infrastructure: [e.g., Kubernetes, bare-metal, laptop]: Kubernetes
 - Deployment tool: [e.g., helm, jsonnet]: 

**Additional Context**

Without digging very deep, these are some guesses of where things could be going wrong:
- misconfigured defaults on the querier worker
- Stopping the pool in parallel with the querier worker (could be that the pool needs to be shut down after the worker) https://github.com/grafana/tempo/blob/3b9f309a9e9a1e0e943ee8b7a3d8c4f458962445/modules/querier/querier.go#L92

**Comment by joe-elliott** [comment]
Queriers can also incorrectly 404 on startup. This appears to be b/c they will answer queries before they have completely polled the blocklist.

**Comment by kvrhdn** [comment]
Okay, I'm seeing a couple of issues related to queriers stopping and starting:

#### A new querier receives a request while its ring is still empty

```
2021-07-09T12:03:00.080690887Z stderr F level=warn ts=2021-07-09T12:03:00.078940012Z caller=querier.go:105 msg=Querier.starting
2021-07-09T12:03:00.080693554Z stderr F level=warn ts=2021-07-09T12:03:00.07954222Z caller=querier.go:107 msg="Querier.starting - exiting"
2021-07-09T12:03:00.080695137Z stderr F level=warn ts=2021-07-09T12:03:00.079652137Z caller=querier.go:120 msg=Querier.running
2021-07-09T12:03:00.205393262Z stderr F level=warn ts=2021-07-09T12:03:00.204881845Z caller=logging.go:71 traceID=cbfd67d312bc08 msg="GET /querier/api/traces/2db629d3a303a2bc03fe2d25d4ea6f90?mode=ingesters (500) 3.665375ms Response: \"error finding ingesters in Querier.FindTraceByID: empty ring\\n\" ws: false; Accept: application/protobuf; Accept-Encoding: gzip; Content-Type: application/protobuf; User-Agent: hey/0.0.1; X-Scope-Orgid: single-tenant; uber-trace-id: 00cbfd67d312bc08:6178558e5275a000:00cbfd67d312bc08:1; "
```

Something similar might happen with the blocklist, but I haven't encountered this locally yet.

Possible solution:

- wait until the ring has at least replication factor instances in `Querier.starting`

#### When a querier is shut down, all requests are immediately cancelled

```
2021-07-09T12:36:19.940548251Z stderr F level=warn ts=2021-07-09T12:36:19.939617001Z caller=querier.go:122 msg="Querier.running - exiting"
2021-07-09T12:36:19.940554709Z stderr F level=warn ts=2021-07-09T12:36:19.939638209Z caller=querier.go:139 msg=Querier.stopping
2021-07-09T12:36:19.944722417Z stderr F level=warn ts=2021-07-09T12:36:19.944370959Z caller=logging.go:71 traceID=1abda046b7c857f2 msg="GET /api/traces/2db629d3a303a2bc03fe2d25d4ea6f90 (500) 559.52475ms Response: \"rpc error: code = Canceled desc = context canceled\\n\" ws: false; Accept-Encoding: gzip; Content-Type: application/protobuf; User-Agent: hey/0.0.1; "
```

[Query sharding in the frontend performs all requests in parallel](https://github.com/grafana/tempo/blob/main/modules/frontend/querysharding.go#L117-L118) and does not retry a cancelled request.

Possible solutions:

- when a querier is shut down, don't cancel in-flight requests but signal the query-frontend to stop sending new requests
- retry cancelled requests in query-frontend (can we extract this from the error?)

---

fyi, to reproduce these issues I'm sending a constant load to the query-frontend while scaling the querier up and down.

My setup:
- tempo in microservices mode
- [hey](https://github.com/rakyll/hey) just spams a single traces call
```
./hey -z 30m -m GET -T application/protobuf http://localhost:3100/api/traces/2db629d3a303a2bc03fe2d25d4ea6f90
```

Because hey is constantly sending queries, at least one or two calls fail while the querier is changing

**Comment by joe-elliott** [comment]
Excellent research. I think we just need a better startup/shutdown procedure for the querier.

On startup we need to wait for the querier to join the cluster, get the ring and poll the blocklist. After these operations are complete it can connect to the query frontend and begin serving queries.

On shutdown we need to stop serving queries, allow all in flight queries to finish and then exit.

Bonus points for adding a retry option on cancelled request to the query-frontend.

**Comment by kvrhdn** [comment]
Implemented retry middleware in the query-frontend: #814. This will avoid failing the entire query while one of the queriers is shutting down.

Queriers cancelling in-flight requests is a known issue in Cortex: https://github.com/cortexproject/cortex/issues/3571#issuecomment-786563209.
I'll take some time to investigate if it's possible to solve this upstream in Cortex.

This retry middleware will also 'solve' the empty ring error, but I still want to get the querier lifecycle right to avoid unnecessary retries. Additionally, an empty blocklist can't be detected with an error.

**Comment by joe-elliott** [comment]
I like the idea of having a configurable # of retries even if we also solve the "cancelling in-flight" issue. nice work!

**Comment by kvrhdn** [comment]
After rolling out #814 I haven't seen any 500s anymore when rolling queriers. I do see an elevated amount of 404s while rolling that seem to originate from traces flushed to the backend (searching the ingesters doesn't cause 404s).

This confirms the issue that the querier accepts request before polling the entire blocklist.

**Comment by joe-elliott** [comment]
@kvrhdn do you mind making new issue for the 404s and we can close this one?
